components/render/atlastexture.cpp

```diff
@@ -0,0 +1,154 @@
+            glClear(GL_COLOR_BUFFER_BIT);
+
+            glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_MIN_FILTER, GL_LINEAR);
+            glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_MAG_FILTER, GL_LINEAR);
+            glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_WRAP_S, GL_CLAMP_TO_EDGE);
+            glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_WRAP_T, GL_CLAMP_TO_EDGE);
+        }
+
+        glBindFramebuffer(GL_DRAW_FRAMEBUFFER, 0);
+        glDeleteFramebuffers(1, &fbo);
+
+#ifdef DEBUG_ATLAS_TEXTURE
+        GLint internalFormat = 0;
+        glGetTexLevelParameteriv(GL_TEXTURE_2D, 0, GL_TEXTURE_INTERNAL_FORMAT, &internalFormat);
+        printf("MaxTextureSize %d, used (%d, %d, %d), 0x%04X\n", maxTextureSize, mTextureWidth, mTextureHeight, mTextureLayers, internalFormat);
+#endif
+    }
+
+    size_t AtlasTexture::addTexture(int32_t width, int32_t height, const void* data)
+    {
+        // Alignment of 4 is required for default pixel store alignment or for using compressed textures.
+        // Padding is useful for avoiding artifacts when highlighting edges in a packed atlas texture.
+        // Only the bottom and right need be padded when adding an image.
+        // e.g. adding two 3x3 images beside one another (I=image, P=padding, A=alignment padding):
+        // PADDING = 1        PADDING = 2            PADDING = 1        PADDING = 2
+        // ALIGNMENT = 1      ALIGNMENT = 1          ALIGNMENT = 4      ALIGNMENT = 4
+        // I I I P I I I P    I I I P P I I I P P    I I I P I I I P    I I I P P A A A I I I P P A A A
+        // I I I P I I I P    I I I P P I I I P P    I I I P I I I P    I I I P P A A A I I I P P A A A
+        // I I I P I I I P    I I I P P I I I P P    I I I P I I I P    I I I P P A A A I I I P P A A A
+        // P P P P P P P P    P P P P P P P P P P    P P P P P P P P    P P P P P A A A P P P P P A A A
+        //                    P P P P P P P P P P                       P P P P P A A A P P P P P A A A
+        //                                                              A A A A A A A A A A A A A A A A
+        //                                                              A A A A A A A A A A A A A A A A
+        //                                                              A A A A A A A A A A A A A A A A
+        static const int32_t PADDING = 1;   // Must be >= 0
+        static const int32_t ALIGNMENT = 1; // Must be >= 1
+        int32_t paddedWidth = (width + PADDING + ALIGNMENT - 1) / ALIGNMENT * ALIGNMENT;
+        int32_t paddedHeight = (height + PADDING + ALIGNMENT - 1) / ALIGNMENT * ALIGNMENT;
+
+        // This assert may cause trouble as one of the Diablo 1 images is 10752 pixels high.
+        // However ~46% (https://feedback.wildfiregames.com/report/opengl/feature/GL_MAX_TEXTURE_SIZE)
+        // of graphics cards are limited to 8192, not quite sure how it ever worked..
+        release_assert(paddedWidth <= mTextureWidth && paddedHeight <= mTextureHeight); // Texture size too small...
```

Comment:
This comment mentions my main portability concern

Reply:
"ui_art/font42g.pcx"
Exported it to jpeg:
![font42g](https://user-images.githubusercontent.com/8636545/54111680-558b3a80-4449-11e9-8dd7-e57ed281c424.jpg)

Reply:
Looks like it does work, using Render::loadResizedSprite. So, problem solved :D

Reply:
Just requires:
- updating `Renderer::loadFonts()` to pass the additional info (`resize=640x672&tileSize=40x42`)
- `SpriteCache::get(const std::string& path)` also needs to handle `pair[0] == "resize"` etc like `SpriteCache::get(uint32_t index)` (can probably make the functionality common to both).
- `PcxFontInfo` needs to be given the extra information ("tile" size, font height (Note: font height 24 actually 26, 30 actually 31), texture width/height).
- `PcxFontInfo::queryGlyph` needs to be updated to return the correct uv coorinates. Note: resizing the font image also transposes the sub images in this case i.e.
    ```
    1         resize ->     1    2    3    4
    2                       5    6    7    8
    3                       9   ....
    4
    ...